frame/election-provider-multi-phase/src/unsigned.rs

```diff
@@ -62,33 +77,151 @@ impl From<FeasibilityError> for MinerError {
 	}
 }
 
+/// Save a given call into OCW storage.
+fn save_solution<T: Config>(call: &Call<T>) -> Result<(), MinerError> {
+	let storage = StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL);
+	match storage.mutate::<_, (), _>(|_| Ok(call.clone())) {
+		Ok(Ok(_)) => Ok(()),
+		Ok(Err(_)) => Err(MinerError::FailedToStoreSolution),
+		Err(_) => {
+			// this branch should be unreachable according to the definition of `StorageValueRef::mutate`:
+			// that function should only ever `Err` if the closure we pass it return an error.
+			// however, for safety in case the definition changes, we do not optimize the branch away
+			// or panic.
+			Err(MinerError::FailedToStoreSolution)
+		},
+	}
+}
+
+/// Get a saved solution from OCW storage if it exists.
+fn restore_solution<T: Config>() -> Result<Call<T>, MinerError> {
+	StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL)
+		.get()
+		.flatten()
+		.ok_or(MinerError::NoStoredSolution)
+}
+
+/// Clear a saved solution from OCW storage.
+pub(super) fn kill_ocw_solution<T: Config>() {
+	let mut storage = StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL);
+	storage.clear();
+}
+
+/// `true` when OCW storage contains a solution
+///
+/// More precise than `restore_solution::<T>().is_ok()`; that invocation will return `false`
+/// if a solution exists but cannot be decoded, whereas this just checks whether an item is present.
+#[cfg(test)]
+fn ocw_solution_exists<T: Config>() -> bool {
+	matches!(
+		StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL).get::<Call<T>>(),
+		Some(_),
+	)
+}
+
 impl<T: Config> Pallet<T> {
-	/// Mine a new solution, and submit it back to the chain as an unsigned transaction.
-	pub fn mine_check_and_submit() -> Result<(), MinerError> {
+	/// Attempt to restore a solution from cache. Otherwise, compute it fresh. Either way, submit
+	/// if our call's score is greater than that of the cached solution.
+	pub fn restore_or_compute_then_maybe_submit() -> Result<(), MinerError> {
+		log!(
+			debug,
+			"OCW attempting to restore or compute an unsigned solution for the current election"
+		);
+
+		let (call, _score) = restore_solution::<T>()
+			.and_then(|call| {
```

Comment:
we never check that the stored solution is correct for the current chain ?
If it is invalid we should mine a new solution. (actually even if it is valid, it might worth mining a new solution considering the nominations might have changed).

Let's say this following scenario:
* Block 10a is produced
* Block 11a is produced and open the election, the validator run the offchain worker and find a solution.
* Block 10b is produced and is the new best block and contains a transaction which changes a nomination (making the solution previously mined invalid), the validator do the reorg and run offchain worker
* Block 11b is produced and open the election, the validator doesn't submit a solution considering that the other solution head is 11, validation will submit a new one at 11+5.
* Block 12b, 13b, 14b, 15b, 16b are produced, validator submit the invalid solution to the queue. and will never mine a valid one.

Am I correct ?

Reply:
> (actually even if it is valid, it might worth mining a new solution considering the nominations might have changed)

At that point we may as well just unconditionally mine a new solution and never store one. We keep narrowing the circumstances in which we can use the cached one.

> we never check that the stored solution is correct for the current chain ?

This seems related to https://github.com/paritytech/substrate/pull/8290#discussion_r605669625. Still, I see what you mean. For now, what about https://github.com/paritytech/substrate/pull/8290/commits/fc71d3dc1b6fa820a885b04346bf0b6cd89628d8 ?

Reply:
> > (actually even if it is valid, it might worth mining a new solution considering the nominations might have changed)
> 
> At that point we may as well just unconditionally mine a new solution and never store one. We keep narrowing the circumstances in which we can use the cached one.

We can also hash the election data alongside the solution and only recompute if the election data has changed (which means there is a reorg on a fork with transactions changing nominations).

That may be fine not to mine a solution again, but I feel like ideally every validator should have mine a solution for the election data of the final chain (and not for the election data of a fork).

> > we never check that the stored solution is correct for the current chain ?
> 
> This seems related to [#8290 (comment)](https://github.com/paritytech/substrate/pull/8290#discussion_r605669625). Still, I see what you mean. For now, what about [fc71d3d](https://github.com/paritytech/substrate/commit/fc71d3dc1b6fa820a885b04346bf0b6cd89628d8) ?

This only includes the validation checks, (as in ValidateUnsigned::validate_unsigned), it doesn't check that the voter and targets are correct. (and voter and targets could have changed in the block 10b in the example)

Reply:
We could run the `feasibility-check` here to ensure a saved solution is always correct, at least up to the current best block. This is also aligned with the ideas in https://github.com/paritytech/substrate/issues/8641. The code would be inefficient but more correct. Similar to `mine_checked`, we pass the solution to both `pre_dispatch_checks` and `feasibility_check`. Then we don't need to check round and score manually either, as it is checked inside those functions.

Reply:
> This is also aligned with the ideas in #8641.

I don't see where the idea above say that the retrieved solution can be wrong because the election data has changed.

> We could run the feasibility-check here to ensure a saved solution is always correct, 

I think we should do so. Or find a way to ensure the election data hasn't change https://github.com/paritytech/substrate/issues/8641#issuecomment-824669149

Reply:
if my above scenario is correct, then we should run the feasibility-check when retrieving the solution.
It is less efficient, but not doing it is incorrect as far as I can see.

Reply:
> I don't see where the idea above say that the retrieved solution can be wrong because the election data has changed.

Indeed, I meant that my suggestion would also help with #8641. but #8641 is about something else. 

I think hashing the snapshot is also nice and a bit more efficient, but I don't have a preference. Calling `feasibility_check` gives us the same guarantee and it is not deadly or anything.

Reply:
8eaf4811e does `Self::feasibility_check` on retrieval. I'd like to have the efficiency pass, with the election data hash etc., be a separate PR.

Reply:
errrr we don't need to check score and round anymore before `feasibility_check` because it is already done as a part of the `feasibility_check`. Also now that we are choosing to (reasonably) sacrifice speed for security we should just to the same things that we do after mining a new solution: call both `feasibility_check` and `pre_dispatch_checks`. Then you also check the `Phase` one last time (we will refactor this as mentioned in #8641)

Reply:
Ah, right. f4b52f036

Reply:
thanks